src/lib/core/datetime/date-adapter.ts

```diff
@@ -0,0 +1,170 @@
+/** Adapts type `D` to be usable as a date by material components that work with dates. */
+export abstract class DateAdapter<D> {
+  /**
+   * Gets the year component of the given date.
+   * @param date The date to extract the year from.
+   * @returns The year component.
+   */
+  abstract getYear(date: D): number;
+
+  /**
+   * Gets the month component of the given date.
+   * @param date The date to extract the month from.
+   * @returns The month component (0-indexed, 0 = January).
+   */
+  abstract getMonth(date: D): number;
+
+  /**
+   * Gets the date of the month component of the given date.
+   * @param date The date to extract the date of the month from.
+   * @returns The month component (1-indexed, 1 = first of month).
+   */
+  abstract getDate(date: D): number;
+
+  /**
+   * Gets the day of the week component of the given date.
+   * @param date The date to extract the day of the week from.
+   * @returns The month component (0-indexed, 0 = Sunday).
+   */
+  abstract getDay(date: D): number;
+
+  /**
+   * Gets a list of names for the months.
+   * @param style The naming style (e.g. long = 'January', short = 'Jan', narrow = 'J').
```

Comment:
I called it `style` because I don't want people confusing it with the `format: any` that they would pass to the `parse` or `format` functions